core/blockchain.go

```diff
@@ -1345,6 +1379,10 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 		if !istanbul.IsFirstBlockOfEpoch(block.NumberU64(), bc.chainConfig.Istanbul.Epoch) {
 			epochNum := istanbul.GetEpochNumber(block.NumberU64(), bc.chainConfig.Istanbul.Epoch)
 
+			if err := bc.checkUptimes(state, epochNum); err != nil {
```

Comment:
I see that you are updating the updates here.  WDYT of updating it when it processes the header instead?  AFAIK, uptime only needs the parent aggregated seal field.  And if the logic is moved over there, then it should work for both full and fast sync.

Reply:
I can check. But it might slow down the fast sync a bit.

Reply:
Ya, you make a good point.

IMO, if we want to continue on with the "spirit" of fast sync, then this node should download the uptime values from other nodes, and those values are "committed" onto the header.  But seems like that's going to be big changes.

Also, if the uptimes are computed here for fast sync, is there a chance that they could still be incorrect when this node reaches the head node and transitions into full sync?  E.g. when the node does reach that point, are we guaranteed to have downloaded all the historical blocks?

Reply:
This will be called only after it has transitioned to full sync, and all the previous blocks should be available.

Reply:
I believe that the way the header verification logic is written inside `istanbul/`, it doesn't have access to the state, so we'd need to break some abstraction in order to make this work.

Also, the uptime calculation on non-epoch blocks, so you'd be adding extra information on the header that `lightest` clients do not need.

For the above two reasons, I do not think we should move this logic to the header verification functions